feat(terraform): add autocompletion for Terraform 0.13

haidaraM opened the pull request:
As of Terraform ~0.12.29~ 0.13.3

## Standards checklist:

- [x] The PR title is descriptive.
- [x] The PR doesn't replicate another PR which is already open.
- [x] I have read the contribution guide and followed all the instructions.
- [x] The code follows the code style guide detailed in the wiki.
- [x] The code is mine or it's from somewhere with an MIT-compatible license.
- [x] The code is efficient, to the best of my ability, and does not waste computer resources.
- [x] The code is stable and I have tested it myself, to the best of my abilities.

## Changes:

- Add autocompletion for login/logout and force-unlock
- Add completion for 0.13upgrade
- Remove legacy push command
- Add -json flag to version command

haidaraM commented:
CC @mcornella :-)

mcornella commented:
Hi @haidaraM, the README now says v0.13, but there was no completion change. Can you verify that for that version, the completion is the same as the one in this PR?

haidaraM commented:
Hi @mcornella 
Just checked the commands for Terraform v0.13 and there are some differences:
- The command `0.12upgrade` is no longer available (https://github.com/hashicorp/terraform/pull/24403). A warning is shown and the command does nothing. Wondering if I should remove it or not from the completion ? I guess lot of people are still on v0.12.
- I added the command `0.13upgrade`. Basically the same as `0.12upgrade`
- The `push` command (not `state push`) is also no longer supported and doesn't work anymore. Looks like it was the case in v0.12 also. I removed it.
- There is a new `-json` flag in the `version` command (https://github.com/hashicorp/terraform/pull/25252). I added it.

There are others changes that I will try to add later: 2 subcommands for `providers` (`schema` and `mirror`), 1 subcommand for `state` (`replace-provider`) and maybe other changes...

mcornella commented:
Is this ready?

haidaraM commented:
Yes, @mcornella

alexjurkiewicz commented on plugins/terraform/_terraform:
```diff
@@ -31,45 +34,51 @@ __012upgrade() {
-        '-var-file=[(foo) Set variables in the Terraform configuration from a file. If "terraform.tfvars" or any ".auto.tfvars" files are present, they will be automatically loaded.]'
+        '*-target=[(resource) Resource to target. Operation will be limited to this resource and its dependencies. This flag can be used multiple times.]:target:__statelist' \
+        '*-var[("foo=bar") Set a variable in the Terraform configuration. This flag can be set multiple times.]' \
+        '*-var-file=[(foo) Set variables in the Terraform configuration from a file. If "terraform.tfvars" or any ".auto.tfvars" files are present, they will be automatically loaded.]:file:_files'
```
```suggestion
        '*-var-file=[(foo) Set variables in the Terraform configuration from a file. If "terraform.tfvars" or any ".auto.tfvars" files are present, they will be automatically loaded.]:file:_files -g "*.tfvars{,.json}"'
```

This will restrict matches to only `.tfvars` and `.tfvars.json` files.

alexjurkiewicz reviewed (approved):
This update works really well! Thanks for writing it 🙌 

It's possible to make some of the file completions more specific.

alexjurkiewicz commented on plugins/terraform/_terraform:
```diff
@@ -31,45 +34,51 @@ __012upgrade() {
-        '-backup=[(path) Path to backup the existing state file before modifying. Defaults to the "-state-out" path with ".backup" extension. Set to "-" to disable backup.]' \
         '-auto-approve[Skip interactive approval of plan before applying.]' \
-        '-lock=[(true) Lock the state file when locking is supported.]' \
+        '-backup=[(path) Path to backup the existing state file before modifying. Defaults to the "-state-out" path with ".backup" extension. Set to "-" to disable backup.]' \
```
```suggestion
        '-backup=[(path) Path to backup the existing state file before modifying. Defaults to the "-state-out" path with ".backup" extension. Set to "-" to disable backup.]:backupfile:_files -g "*.backup"' \
```

haidaraM replied:
Done

alexjurkiewicz commented on plugins/terraform/_terraform:
```diff
@@ -31,45 +34,51 @@ __012upgrade() {
         '-no-color[If specified, output wil be colorless.]' \
         '-parallelism=[(10) Limit the number of parallel resource operations.]' \
         '-refresh=[(true) Update state prior to checking for differences. This has no effect if a plan file is given to apply.]' \
         '-state=[(terraform.tfstate) Path to read and save state (unless state-out is specified).]' \
```
```suggestion
        '-state=[(terraform.tfstate) Path to read and save state (unless state-out is specified).]:statefile:_files -g "*.tfstate"' \
```

Also applies to `-state-out` args.

haidaraM commented:
@alexjurkiewicz Just applied the changes :-)

mcornella commented:
Thanks!
